src/sink/kafka.rs

```diff
@@ -0,0 +1,214 @@
+use rdkafka::client::EmptyContext;
+use rdkafka::config::ClientConfig;
+use rdkafka::error::{KafkaError, RDKafkaError};
+use rdkafka::message::{Message, OwnedMessage};
+use rdkafka::producer::FutureProducer;
+use rdkafka::producer::future_producer::DeliveryFuture;
+use sink::Sink;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use util::Valve;
+
+lazy_static! {
+    /// Total records published.
+    pub static ref KAFKA_PUBLISH_SUCCESS_SUM: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
+    /// Total record publish retries.
+    pub static ref KAFKA_PUBLISH_RETRY_SUM: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
+    /// Total record publish failures.
+    pub static ref KAFKA_PUBLISH_FAILURE_SUM: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
+}
+
+/// Config options for Kafka config.
+#[derive(Clone, Debug, Deserialize)]
+pub struct KafkaConfig {
+    /// Canonical name for the given Kafka sink.
+    pub config_path: Option<String>,
+    /// Kafka topic to publish to.
+    pub topic_name: Option<String>,
+    /// Kafka brokers. This is a comma-separated list of host or host:port.
+    pub brokers: Option<String>,
+    /// Underlying librdkafka configuration.
+    pub rdkafka_config: Option<HashMap<String, String>>,
+    /// How often (seconds) the in-flight messages are checked for delivery.
+    /// Default = 1 second
+    pub flush_interval: u64,
+}
+
+impl Default for KafkaConfig {
+    fn default() -> KafkaConfig {
+        KafkaConfig {
+            config_path: None,
+            topic_name: None,
+            brokers: None,
+            rdkafka_config: None,
+            flush_interval: 1,
+        }
+    }
+}
+
+/// Kafka sink internal state.
+pub struct Kafka {
+    /// Name of the stream we are publishing to.
+    topic_name: String,
+    /// A message producers.
+    producer: FutureProducer<EmptyContext>,
+    // In-flight messages.
+    messages: Vec<DeliveryFuture>,
+    /// How often (seconds) the in-flight messages are checked for delivery.
+    flush_interval: u64,
+}
+
+impl Sink<KafkaConfig> for Kafka {
+    fn init(config: KafkaConfig) -> Self {
+        if config.topic_name.is_none() {
+            panic!("No Kafka topic name provided!");
+        }
+        if config.brokers.is_none() {
+            panic!("No Kafka brokers provided!")
+        }
+
+        let mut producer_config = ClientConfig::new();
+        if let Some(ref map) = config.rdkafka_config {
+            for (key, value) in map.iter() {
+                producer_config.set(key, value);
+            }
+        }
+        producer_config.set("bootstrap.servers", &config.brokers.unwrap()[..]);
+
+        Kafka {
+            topic_name: config.topic_name.unwrap(),
+            producer: producer_config.create::<FutureProducer<_>>().unwrap(),
+            messages: Vec::new(),
+            flush_interval: config.flush_interval,
+        }
+    }
+
+    fn valve_state(&self) -> Valve {
```

Comment:
We may want to have some kind of valve. I forget what the overhead of a Future is but we're buffering them up into `self.messages`, exposing us to a potential unbounded memory buffer. 

Which, fancy trick, buffering Futures vs the payloads.

Reply:
I _think_ the flush function acts as a natural valve already to protect against a crazy amount of payloads already. In flush, each future is `wait`-ed upon, and by the time we return, only messages that failed and are now in-flight for retry will remain in the buffer.

Reply:
Ah ha, I missed the wait.

Reply:
Latest commit adds valve control based on the number of bytes that are currently in flight

Reply:
Oh, interesting choice! Elsewhere in cernan the 'payload' is the standard metric, which was chosen vs. byte count arbitrarily. I guess if someone is willing to lose 100mb but has very tiny messages they may lose some, but an Ops person -- on the flip side -- will be able to guarantee the memory constraint. 

Hmm, makes me wonder now if we oughtn't to allow byte restrictions elsewhere, too.